org.eclipse.languageserver/src/org/eclipse/languageserver/operations/format/LSPFormatHandler.java

```diff
@@ -49,31 +55,61 @@ public Object execute(ExecutionEvent event) throws ExecutionException {
 			if (info != null) {
 				ISelection sel = ((AbstractTextEditor) part).getSelectionProvider().getSelection();
 				if (sel instanceof TextSelection) {
+					final Shell shell = HandlerUtil.getActiveShell(event);
 					DocumentFormattingParams params = new DocumentFormattingParamsBuilder()
-							.textDocument(info.getFileUri().toString())
-							.build();
+					        .textDocument(info.getFileUri().toString())
+					        .options(new FormattingOptionsBuilder().build())
+					        .build();
 					CompletableFuture<List<? extends TextEdit>> formatter = info.getLanguageClient().getTextDocumentService().formatting(params);
 					formatter.thenAccept((List<? extends TextEdit> t) -> {
-						for (TextEdit textEdit : t) {
-							try {
-								LSPEclipseUtils.applyEdit(textEdit, document);
-							} catch (BadLocationException e) {
-								// TODO Auto-generated catch block
-								e.printStackTrace();
-							}
-						}
+						shell.getDisplay().asyncExec(() -> {
+							applyChanges(info.getDocument(), t);
+						});
 					});
 				}
 			}
 		}
 		return null;
 	}
 
+	private void applyChanges(IDocument document, List<? extends TextEdit> t) {
```

Comment:
This cool utils method should be moved to LSPEclipseUtils as multiple LSP operation can take advantage of it.
Also, could it happen in a non UI-Thread (just asking there, I really don't know)